source/MRViewer/MRSurfaceManipulationWidget.cpp

```diff
@@ -0,0 +1,266 @@
+    regionOld_ = VertBitSet( numV, false );
+    changedValues_ = VertScalars( numV, 0.f );
+    distances_ = VertScalars( numV, 0.f );
+
+    obj_->setAncillaryTexture( { { { Color { 255, 64, 64, 255 }, Color { 0, 0, 0, 0 } }, Vector2i { 1, 2 } } } );
+    uvs_ = VertUVCoords( numV, { 0, 1 } );
+    obj_->setAncillaryUVCoords( uvs_ );
+}
+
+void SurfaceManipulationWidget::reset()
+{
+    obj_->clearAncillaryTexture();
+    obj_.reset();
+
+    region_ = {};
+    regionExpanded_ = {};
+    regionOld_ = {};
+    changedValues_ = {};
+    distances_ = {};
+
+    uvs_ = {};
+
+    changeMeshAction_.reset();
+}
+
+void SurfaceManipulationWidget::setSettings( const Settings& settings )
+{
+    settings_ = settings;
+    settings_.radius = std::max( settings_.radius, 1.f );
+    settings_.force = std::clamp( settings_.force, 1.f, 100.f );
+    settings_.intensity = std::clamp( settings_.intensity, 1.f, 100.f );
+}
+
+bool SurfaceManipulationWidget::onMouseDown( Viewer::MouseButton button, int /*modifier*/ )
+{
+    if ( button != MouseButton::Left )
+        return false;
+
+    auto [obj, pick] = getViewerInstance().viewport().pick_render_object();
+    if ( !obj || obj != obj_ )
+        return false;
+
+    changeMeshAction_ = std::make_shared<ChangeMeshAction>( "Change mesh surface", obj_ );
+    mousePressed_ = true;
+
+    return true;
+}
+
+bool SurfaceManipulationWidget::onMouseUp( Viewer::MouseButton button, int /*modifier*/ )
+{
+    if ( button != MouseButton::Left )
+        return false;
+
+    mousePressed_ = false;
+
+    AppendHistory( changeMeshAction_ );
+    changeMeshAction_.reset();
+
+    for ( auto v : regionOld_ )
+        changedValues_[v] = 0.f;
+
+    return true;
+}
+
+bool SurfaceManipulationWidget::onMouseMove( int mouse_x, int mouse_y )
+{
+    updateRegion_( Vector2f{ float( mouse_x ), float( mouse_y ) } );
+
+    return true;
+}
+
+bool SurfaceManipulationWidget::onKeyDown( int /*key*/, int modifier )
+{
+    bool res = false;
+    if ( modifier & GLFW_MOD_SHIFT )
+    {
+        onlySmooth_ = true;
+        res = true;
+    }
+    if ( modifier & GLFW_MOD_CONTROL )
+    {
+        direction_ = -1.f;
+        res = true;
+    }
+
+    return res;
+}
+
+bool SurfaceManipulationWidget::onKeyUp( int /*key*/, int modifier )
+{
+    bool res = false;
+    if ( !( modifier & GLFW_MOD_SHIFT ) )
+    {
+        onlySmooth_ = false;
+        res = true;
+    }
+    if ( !( modifier & GLFW_MOD_CONTROL ) )
+    {
+        direction_ = 1.f;
+        res = true;
+    }
+
+    return res;
+}
+
+void SurfaceManipulationWidget::postDraw()
+{
+    if ( mousePressed_ )
+        changeSurface_();
+    
+    updateRegion_( mousePos_ );
+}
+
+
+void SurfaceManipulationWidget::changeSurface_()
+{
+    if ( !region_.any() )
+        return;
+
+    MR_TIMER
+
+    if ( onlySmooth_ )
+    {
+        positionVertsSmoothly( *obj_->varMesh(), region_, Laplacian::EdgeWeights::Unit );
+        obj_->setDirtyFlags( DIRTY_PRIMITIVES );
```

Comment:
better use `relax` function here, and  `DIRTY_POSITIONS` instead of `DIRTY_PRIMITIVES`

Reply:
fixed